File: src/handlers/users/provider/github.rs
use crate::{
    config::Config,
    handlers::users::provider::{Provider, UserProvidedData},
};
use reqwest::{Client, Error};
use serde::Deserialize;

#[derive(Deserialize)]
struct PictureData {
    pub is_silhouette: bool,
    pub url: String,
}

#[derive(Deserialize)]
struct Data {
    pub data: PictureData,
}

#[derive(Deserialize)]
struct GithubUser {
    pub email: Option<String>,
    pub name: Option<String>,
    pub avatar_url: Option<String>,
}

#[derive(Deserialize)]
struct GithubUserEmail {
    pub email: String,
    pub primary: bool,
    pub verified: bool,
}

#[derive(Clone)]
pub struct GithubProvider {
    config: Config,
}

impl GithubProvider {
    pub fn new(config: Config) -> GithubProvider {
        return GithubProvider { config };
    }
}

impl Provider for GithubProvider {
    fn enabled(&self) -> bool {
        return self.config.github_enabled;
    }

    fn auth_url(&self) -> String {
        return String::from("https://github.com/login/oauth/authorize");
    }

    fn token_url(&self) -> String {
        return String::from("https://github.com/login/oauth/access_token");
    }

    fn client_id(&self) -> String {
        return self.config.github_client_id.clone().unwrap();
    }

    fn client_secret(&self) -> String {
        return self.config.github_client_secret.clone().unwrap();
    }

    fn scopes(&self) -> Vec<String> {
        return vec![String::from("user:email")];
    }

    fn get_user_data(&self, access_token: String) -> Result<UserProvidedData, Error> {
        let client = Client::new();

        let mut response = client.get("https://api.github.com/user").bearer_auth(access_token.clone()).send()?;

        let user: GithubUser = response.json()?;

        response = client.get("https://api.github.com/user/emails").bearer_auth(access_token).send()?;

        let emails: Vec<GithubUserEmail> = response.json()?;

        let mut data = UserProvidedData {
            verified: false,
            email: None,
            name: user.name,
            avatar: user.avatar_url,
        };

        for email in emails {
            if email.primary {
                data.email = Some(email.email);
                data.verified = email.verified;
                break;
            }
        }

        return Ok(data);
    }
}

File: handlers/provider/github.go
package provider

import (
	"encoding/json"
	"fmt"
	"net/http"

	"github.com/zolamk/trust/config"
	"golang.org/x/oauth2"
	"golang.org/x/oauth2/github"
)

type githubUser struct {
	Name      *string `json:"name"`
	Email     *string `json:"email"`
	AvatarURL *string `json:"avatar_url"`
}

type githubUserEmail struct {
	Email    string
	Primary  bool
	Verified bool
}

type GithubProvider struct {
	c *config.Config
}

func (g *GithubProvider) name() string {
	return "github"
}

func (g *GithubProvider) enabled() bool {
	return g.c.Github.Enabled
}

func (g *GithubProvider) get_config() *oauth2.Config {
	return &oauth2.Config{
		ClientID:     g.c.Github.ID,
		ClientSecret: g.c.Github.Secret,
		Endpoint:     github.Endpoint,
		Scopes:       []string{"user:email"},
		RedirectURL:  fmt.Sprintf("%s/authorize/callback", g.c.InstanceURL),
	}
}

func (g *GithubProvider) get_user_data(access_token string) (*UserData, error) {

	req, err := http.NewRequest("GET", "https://api.github.com/user", nil)

	if err != nil {
		return nil, err
	}

	req.Header.Add("authorization", fmt.Sprintf("Bearer %s", access_token))

	res, err := http.DefaultClient.Do(req)

	if err != nil {
		return nil, err
	}

	decoder := json.NewDecoder(res.Body)

	github_user := &githubUser{}

	if err := decoder.Decode(github_user); err != nil {
		return nil, err
	}

	req, err = http.NewRequest("GET", "https://api.github.com/user/emails", nil)

	if err != nil {
		return nil, err
	}

	req.Header.Add("authorization", fmt.Sprintf("Bearer %s", access_token))

	res, err = http.DefaultClient.Do(req)

	if err != nil {
		return nil, err
	}

	decoder = json.NewDecoder(res.Body)

	github_user_emails := []*githubUserEmail{}

	if err = decoder.Decode(&github_user_emails); err != nil {
		return nil, err
	}

	data := &UserData{
		Name:     github_user.Name,
		Verified: false,
		Avatar:   github_user.AvatarURL,
	}

	for _, email := range github_user_emails {
		if email.Primary {
			data.Email = &email.Email
			data.Verified = email.Verified
			break
		}
	}

	return data, nil

}
